src/core/service.c

```diff
@@ -2758,6 +2761,59 @@ static void service_run_next_main(Service *s) {
         (void) service_set_main_pidref(s, TAKE_PIDREF(pidref), &s->main_command->exec_status.start_timestamp);
 }
 
+static bool service_should_reload_extensions(Unit *u) {
+        Service *s = ASSERT_PTR(SERVICE(u));
+        int r;
+
+        /* Only support this for notify-reload service types. */
+        if (s->type != SERVICE_NOTIFY_RELOAD)
+                return false;
+
+        /* TODO: Add support for user services, which can use
+         * ExtensionDirectories= + notify-reload. For now, skip for user
+         * services. */
+        if (u->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
+                log_unit_debug(UNIT(s), "Not reloading extensions for user services.");
+                return false;
+        }
+
+        r = exec_context_has_vpicked_extensions(&s->exec_context);
+        if (r < 0) {
+                log_unit_warning_errno(u, r, "Failed to determine if service should reload extensions, assuming false: %m");
+                return false;
+        }
+        return r > 0;
+}
+
+static int service_reload_extensions(Unit *u) {
+        /* TODO: do this asynchronously */
```

Comment:
sorry, but this is not acceptable as is. mounting stuff is slow, might involve IPC and whatnot. this *must* be done asynchronously or not at all.

please introduce a proper service substate for this, SERVICE_RELOAD_EXTENSIONS or so and fork() off a worker process that is properly tracked for this, and wait for it as part of the service state engine.

Also, please use unit_fork_helper_process() to fork off the helper.

And set a timer to ensure this is bounded.

you might take inspiration from the socket_chown() logic in socket.c which also does a unit-bound asynchronous operation in a child process.

Reply:
As mentioned below, I will take care of this later